src/main/java/ch/njol/skript/lang/function/FunctionRegistry.java

```diff
@@ -0,0 +1,682 @@
+		}
+
+		// since we are getting a namespace and then updating it by putting it back in the map,
+		// avoid race conditions by ensuring only one thread can access the namespaces map at a time.
+		synchronized (namespaces) {
+			Namespace ns = namespaces.computeIfAbsent(namespaceId, n -> new Namespace());
+
+			FunctionIdentifier identifier = FunctionIdentifier.of(signature);
+
+			// register
+			Set<FunctionIdentifier> identifiersWithName = ns.identifiers.computeIfAbsent(identifier.name, s -> new HashSet<>());
+			boolean exists = identifiersWithName.add(identifier);
+			if (!exists) {
+				alreadyRegisteredError(signature.getName(), identifier, namespaceId);
+			}
+
+			ns.signatures.put(identifier, signature);
+		}
+	}
+
+	/**
+	 * Registers a global function.
+	 *
+	 * @param function The function to register.
+	 * @throws SkriptAPIException if the function name is invalid or if
+	 *                            a function with the same name and parameters is already registered
+	 *                            in this namespace.
+	 */
+	public void register(@NotNull Function<?> function) {
+		register(null, function);
+	}
+
+	/**
+	 * Registers a function.
+	 *
+	 * @param namespace The namespace to register the function in.
+	 *                  If namespace is null, will register this function globally.
+	 *                  Usually represents the path of the script this function is registered in.
+	 * @param function  The function to register.
+	 * @throws SkriptAPIException if the function name is invalid or if
+	 *                            a function with the same name and parameters is already registered
+	 *                            in this namespace.
+	 */
+	public void register(@Nullable String namespace, @NotNull Function<?> function) {
+		Preconditions.checkNotNull(function, "function cannot be null");
+		Skript.debug("Registering function '%s'", function.getName());
+
+		String name = function.getName();
+		if (!name.matches(FUNCTION_NAME_PATTERN)) {
+			throw new SkriptAPIException("Invalid function name '" + name + "'");
+		}
+
+		// namespace
+		NamespaceIdentifier namespaceId;
+		if (namespace != null && function.getSignature().isLocal()) {
+			namespaceId = new NamespaceIdentifier(namespace);
+		} else {
+			namespaceId = GLOBAL_NAMESPACE;
+		}
+
+		FunctionIdentifier identifier = FunctionIdentifier.of(function.getSignature());
+		if (!signatureExists(namespaceId, identifier)) {
+			register(namespace, function.getSignature());
+		}
+
+		// since we are getting a namespace and then updating it by putting it back in the map,
+		// avoid race conditions by ensuring only one thread can access the namespaces map at a time.
+		synchronized (namespaces) {
+			Namespace ns = namespaces.computeIfAbsent(namespaceId, n -> new Namespace());
+
+			Function<?> existing = ns.functions.put(identifier, function);
+			if (existing != null) {
+				alreadyRegisteredError(name, identifier, namespaceId);
+			}
```

Comment:
Do we need to be concerned about the existing function within the map being overridden? The call will replace the existing value